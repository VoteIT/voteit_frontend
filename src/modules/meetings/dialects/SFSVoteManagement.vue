<template>
  <div v-if="groupVotes || assignedVotes">
    <span :class="{ 'text-warning': !allAssigned }">{{ assignedVotes }}/{{ groupVotes }}</span>
    <DefaultDialog v-if="canAssignVotes" v-model="editing">
      <template #activator="{ props }">
        <v-btn size="small" color="accent" class="ml-2" v-bind="props">
          Fördela röster
        </v-btn>
      </template>
      <UserList :user-ids="editUserIds">
        <template #appendItem="{ user }">
          <v-text-field
            variant="outlined"
            density="compact"
            @keydown.stop
            type="number"
            :min="0"
            :max="(editUserVotes.get(user) ?? 0) + editUnassignedVotes"
            :model-value="editUserVotes.get(user)"
            @update:model-value="editUserVotes.set(user, Number($event))"
            :rules="[rules.max((editUserVotes.get(user) ?? 0) + editUnassignedVotes)]"
          />
        </template>
      </UserList>
      <v-alert class="mb-2" :color="editUnassignedVotes ? 'secondary' : undefined">
        {{ editUnassignedVotes }} röster kvar att fördela
      </v-alert>
      <v-alert v-if="errorMessage" :text="errorMessage" type="error" class="mb-2" />
      <div class="text-right">
        <v-btn variant="text" :disabled="working" @click="editing = false">
          {{ t('cancel') }}
        </v-btn>
        <v-btn color="primary" :loading="working" :disabled="!!editUnassignedVotes" @click="saveUserVotes">
          {{ t('save') }}
        </v-btn>
      </div>
    </DefaultDialog>
  </div>
</template>

<script lang="ts" setup>
import { sum } from 'itertools'
import { computed, PropType, reactive, ref, watch } from 'vue'
import { useI18n } from 'vue-i18n'

import { socket } from '@/utils/Socket'
import UserList from '@/components/UserList.vue'
import DefaultDialog from '@/components/DefaultDialog.vue'
import { user } from '@/composables/useAuthentication'
import useErrorHandler from '@/composables/useErrorHandler'
import useRules from '@/composables/useRules'

import { GroupMembership, MeetingGroup } from '../types'
import useMeetingGroups from '../useMeetingGroups'
import useMeeting from '../useMeeting'
import { isFinishedMeeting } from '../rules'

const props = defineProps({
  group: {
    type: Object as PropType<MeetingGroup & { memberships: GroupMembership[] }>,
    required: true
  }
})

const { t } = useI18n()
const { isModerator, meetingId } = useMeeting()
const { groupRoles } = useMeetingGroups(meetingId)
const rules = useRules(t)
const { errorMessage, clearErrors, handleSocketError } = useErrorHandler()

interface RoleMembership extends GroupMembership { role: number }
function isRoleMembership (membership: GroupMembership): membership is RoleMembership {
  return !!membership.role
}
const roleMemberships = computed(() => props.group.memberships.filter(isRoleMembership))

const groupVotes = computed(() => props.group.votes || 0)
const assignedVotes = computed(() => {
  return roleMemberships.value
    .reduce((acc, member) => acc + (member.votes ?? 0), 0)
})

const allAssigned = computed(() => props.group.votes === assignedVotes.value)
const leaderRoleId = computed(() => groupRoles.value.find(g => g.role_id === 'leader')?.pk)

/**
 * Meeting is in active state (upcoming, ongoing), and user is moderator or group manager,
 */
const canAssignVotes = computed(() => {
  if (!user.value || !props.group.votes || !roleMemberships.value.length) return false
  return (
    !isFinishedMeeting(meetingId.value) &&
    (
      !isModerator.value ||
      props.group.memberships.some(member => member.user === user.value?.pk && member.role === leaderRoleId.value)
    )
  )
})

// For management modal
const editing = ref(false)
const editUserIds = computed(() => roleMemberships.value.map(member => member.user))
const editUserVotes = reactive(new Map(roleMemberships.value.map(({ user, votes }) => [user, votes || 0])))
const editUnassignedVotes = computed(() => groupVotes.value - sum(editUserVotes.values()))

// Make sure we have the latest values when editing
watch(editing, value => {
  if (!value) return clearErrors()
  for (const { user, votes } of roleMemberships.value) {
    editUserVotes.set(user, votes ?? 0)
  }
  // Remove any users with no current roles
  for (const { user } of props.group.memberships.filter(({ role }) => !role)) {
    editUserVotes.delete(user)
  }
})

const working = ref(false)
async function saveUserVotes () {
  if (editUnassignedVotes.value) throw new Error('Cannot save user votes, becuase not all votes assigned')
  clearErrors()
  working.value = true
  try {
    await socket.call('sfs.set_delegation_voters', {
      meeting_group: props.group.pk,
      weights: [...editUserVotes.entries()]
        .map(([user, weight]) => ({ user, weight }))
        .filter(({ weight }) => weight)
    })
    editing.value = false
  } catch (e) {
    handleSocketError(e)
  }
  working.value = false
}
</script>
