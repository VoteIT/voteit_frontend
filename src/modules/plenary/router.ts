import Plenary from './Plenary.vue'

export default {
  path: '/p/:id/:roomId/:aid',
  name: 'Plenary',
  component: Plenary
}
