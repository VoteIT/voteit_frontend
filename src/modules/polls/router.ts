import { RouterView } from 'vue-router'

import Poll from './PollView.vue'
import Polls from './PollsView.vue'
import StartPoll from './StartPollView.vue'

export default {
  path: 'polls',
  component: RouterView,
  children: [
    {
      path: '',
      name: 'polls',
      component: Polls
    },
    {
      path: ':pid/:pslug',
      name: 'poll',
      component: Poll
    },
    {
      path: 'new',
      name: 'start-poll',
      component: StartPoll
    },
    {
      path: 'new/:aid',
      name: 'start-poll-ai',
      component: StartPoll
    },
    {
      path: ':state',
      name: 'poll-state',
      component: Polls
    }
  ]
}
