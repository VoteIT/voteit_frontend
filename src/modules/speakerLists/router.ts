import SpeakerListsView from './SpeakerListsView.vue'

export default {
  path: 'lists/:system/:aid',
  name: 'speakerLists',
  component: SpeakerListsView
}
