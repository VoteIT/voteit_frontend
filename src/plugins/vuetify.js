import '@mdi/font/css/materialdesignicons.css'
import 'vuetify/lib/styles/main.sass'
import 'vuetify/lib/styles/styles.sass'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/lib/components'
import * as directives from 'vuetify/lib/directives'
import colors from 'vuetify/lib/util/colors'

const light = {
  dark: false,
  variables: {
    'border-color': '223,207,200',
    'border-opacity': 1
  },
  colors: {
    background: '#f8f4f2',
    'app-bar': '#563661',
    'app-bar-divider': '#765681',
    'app-bar-active': '#7976B7',
    'menu-active': '#9C8afa',
    surface: '#ffffff',
    'surface-active': colors.green.lighten5,
    primary: '#3729d8',
    secondary: colors.grey.darken1,
    'secondary-lighten-2': colors.grey.lighten2,
    accent: colors.indigo.darken4,
    error: colors.deepOrange.base,
    info: colors.lightBlue.base,
    'info-lighten-2': colors.lightBlue.lighten2,
    success: '#6ce498',
    'success-lighten-2': '#cbf6db',
    'success-darken-2': '#1da475',
    warning: colors.red.darken2
  }
}

/* Some day
const dark = {
  dark: true,
  variables: {},
  colors: {
    background: '#111',
    surface: '#1e1e1e',
    primary: colors.indigo.darken2,
    secondary: colors.blueGrey.darken2,
    accent: colors.indigo.base,
    error: colors.deepOrange.darken2,
    info: colors.lightBlue.darken2,
    success: colors.green.darken2,
    warning: colors.red.darken4
  }
}
*/

export default createVuetify({
  components,
  directives,
  theme: {
    defaultTheme: 'light',
    themes: {
      light
      // dark
    }
  }
})
