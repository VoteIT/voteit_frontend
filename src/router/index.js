import { createRouter, createWebHistory } from 'vue-router'
import Home from '../views/Home.vue'
import JoinMeeting from '../views/JoinMeeting.vue'
import meeting from './meeting'

const routes = [
  {
    path: '/',
    name: 'Home',
    component: Home
  },
  meeting,
  {
    path: '/join/:id/:slug',
    name: 'JoinMeeting',
    component: JoinMeeting
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
})

export default router
