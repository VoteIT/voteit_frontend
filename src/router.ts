import { createRouter, createWebHistory } from 'vue-router'

import Home from '@/views/Home.vue'
import meeting from '@/modules/meetings/router'
import JoinMeeting from '@/modules/meetings/JoinView.vue'
import plenary from '@/modules/plenary/router'
import speakerFullscreen from '@/modules/speakerLists/fullscreenRouter'
import AppBar from '@/components/AppBar.vue'

const routes = [
  {
    path: '/',
    name: 'Home',
    components: {
      default: Home,
      appBar: AppBar
    }
  },
  meeting,
  plenary,
  speakerFullscreen,
  // Join has url outside meeting, so users don't need meeting roles to visit this view.
  {
    path: '/join/:id/:slug',
    name: 'JoinMeeting',
    components: {
      default: JoinMeeting,
      appBar: AppBar
    }
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
})

export default router
