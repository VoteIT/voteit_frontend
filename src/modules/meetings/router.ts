import AgendaEditView from '../agendas/AgendaEditView.vue'
import AgendaItemView from '../agendas/AgendaItemView.vue'
import polls from '../polls/router'

import ControlPanelView from './ControlPanel.vue'
import ElectoralRegistersView from './electoralRegisters/ElectoralRegistersView.vue'
import JoinMeeting from './JoinView.vue'
import ParticipantsView from './ParticipantsView.vue'
import StartView from './StartView.vue'
import MeetingView from './MeetingView.vue'
import MinutesView from './MinutesView.vue'

export default [
  {
    path: '/m/:id/:slug',
    name: 'MeetingRouterView',
    component: MeetingView,
    children: [
      {
        path: '',
        name: 'meeting',
        component: StartView
      },
      {
        path: 'settings',
        name: 'settings',
        component: ControlPanelView,
        children: [
          {
            path: ':panel',
            name: 'controlPanel',
            component: ControlPanelView
          }
        ]
      },
      {
        path: 'agenda',
        name: 'agendaEdit',
        component: AgendaEditView
      },
      {
        path: 'p',
        name: 'participants',
        component: ParticipantsView,
        children: [
          {
            path: ':tabId',
            name: 'participantsTab',
            component: ParticipantsView
          }
        ]
      },
      {
        path: 'er',
        name: 'electoralRegisters',
        component: ElectoralRegistersView
      },
      {
        path: 'minutes',
        name: 'meetingMinutes',
        component: MinutesView
      },
      polls,
      {
        path: 'a/:aid/:aslug',
        name: 'agendaItem',
        component: AgendaItemView
      }
    ]
  },
  // Join has url outside meeting, so users don't need meeting roles to visit this view.
  {
    path: '/join/:id/:slug',
    name: 'meeting:join',
    component: JoinMeeting
  }
]
