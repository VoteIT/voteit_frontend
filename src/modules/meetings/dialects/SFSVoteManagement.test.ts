import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import { createI18n } from 'vue-i18n'
import router from '@/router'

import SFSVoteManagement from './SFSVoteManagement.vue'
import vuetify from '@/plugins/vuetify'

test('Mount component', () => {
  expect(SFSVoteManagement).toBeTruthy()
  const i18n = createI18n({
    legacy: false,
    messages: { en: {} }
  })
  const wrapper = mount(
    SFSVoteManagement,
    {
      global: {
        plugins: [i18n, router, vuetify]
      },
      props: {
        group: {
          votes: 123,
          memberships: [{ votes: 3 }]
        }
      }
    }
  )
  expect(wrapper.text()).toContain('3/123')
})
