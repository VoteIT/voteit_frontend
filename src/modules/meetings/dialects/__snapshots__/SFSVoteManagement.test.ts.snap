// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mount component 1`] = `
"<div><span class="text-warning">5/123</span>
  <!--v-if-->
</div>"
`;
